### ISSUE TITLE
pgbackrest is expiring diff backups WAL and make them not usable for PITR 

### ISSUE THREAD

**Issue by melrifa1** [pgbackrest/pgbackrest#748]
Please provide the following information when submitting an issue (feature requests or general comments can skip this):

1. pgBackRest version:
pgBackRest 2.10
2. PostgreSQL version:
10.8
3. Operating system/version - if you have more than one server (for example, a database server, a repository host server, one or more standbys), please specify each:
Debian Stretch
4. Did you install pgBackRest from source or from a package?
Source
5. Please attach the following as applicable:
    - `pgbackrest.conf` file(s)
    - `postgresql.conf` settings applicable to pgBackRest (`archive_command`, `archive_mode`, `listen_addresses`, `max_wal_senders`, `wal_level`, `port`)
    - errors in the postgresql log file before or during the time you experienced the issue
    - log file in `/var/log/pgbackrest` for the commands run (e.g. `/var/log/pgbackrest/mystanza_backup.log`)

pgbackrest.conf

```
[global]
log-level-console=warn
log-level-file=info
log-level-stderr=warn
repo-s3-bucket=<redacted>
repo-s3-endpoint=<redacted>
repo-s3-key=<redacted>
repo-s3-key-secret=<redacted>
repo-s3-region=<redacted>
repo-type=s3
retention-full=2000
retention-diff=2000
start-fast=y
backup-standby=n
repo-path=/my_repo_path
process-max=4
recovery-option=standby_mode=on

[mydb]
db-path=/data/pg/10
db-port=5432
```

archive command:
/usr/bin/pgbackrest --config /conf/pgbackrest.conf --stanza=mydb archive-push %p >> /dev/null 2>&1

7. Describe the issue:

I am trying to use PITR and configuring pgbackrest to retain WAL for latest 7 (diff or full) backups
```/usr/bin/pgbackrest --config=/conf/postgresql//pgbackrest.conf --log-level-console=detail --log-level-file=off --log-level-stderr=warn --repo1-path=/postgresql_backup_melrifa2-pg-06-16-10-53 --repo1-retention-archive=7 --repo1-retention-archive-type=diff --repo1-retention-diff=2000 --repo1-retention-full=6 --repo1-s3-bucket=melrifa2-pg-06-16-10-53 --repo1-s3-endpoint=s3-api.us-geo.objectstorage.service.networklayer.com --repo1-s3-key=<redacted> --repo1-s3-key-secret=<redacted> --repo1-s3-region=us-standard --repo1-type=s3 --stanza=formation```
Important piece in the above `--repo1-retention-archive=7 --repo1-retention-archive-type=diff --repo1-retention-full=6 expire`

The result is only the latest backup (diff or full) is usable for PITR and the above command is expiring WAL for all the (diff or full) backups older than the latest.

**Steps to reproduce:**

1- Starting with a full backup
2- Add some data and switch the WAL
3- Take a diff backup with the following command
```2019-06-17 15:15:59.956 P00   INFO: backup command begin 2.10: --no-backup-standby --compress-level=1 --config=/conf/postgresql/pgbackrest.conf --log-level-console=info --log-level-file=off --log-level-stderr=warn --pg1-path=/data/postgresql/10 --pg1-port=5432 --process-max=2 --repo1-path=/postgresql_backup_melrifa2-pg-06-16-10-53 --repo1-retention-diff=2000 --repo1-retention-full=2000 --repo1-s3-bucket=melrifa2-pg-06-16-10-53 --repo1-s3-endpoint=s3-api.us-geo.objectstorage.service.networklayer.com --repo1-s3-key=<redacted> --repo1-s3-key-secret=<redacted> --repo1-s3-region=us-standard --repo1-type=s3 --stanza=formation --start-fast --type=diff```

4- Rotate the backup and WALs produce the following
```
2019-06-17 15:16:48.940 P00   INFO: expire command begin 2.10: --config=/conf/postgresql//pgbackrest.conf --log-level-console=detail --log-level-file=off --log-level-stderr=warn --repo1-path=/postgresql_backup_melrifa2-pg-06-16-10-53 --repo1-retention-archive=7 --repo1-retention-archive-type=diff --repo1-retention-diff=2000 --repo1-retention-full=6 --repo1-s3-bucket=melrifa2-pg-06-16-10-53 --repo1-s3-endpoint=s3-api.us-geo.objectstorage.service.networklayer.com --repo1-s3-key=<redacted> --repo1-s3-key-secret=<redacted> --repo1-s3-region=us-standard --repo1-type=s3 --stanza=formation
2019-06-17 15:16:52.458 P00 DETAIL: archive retention on backup 20190617-150157F, archiveId = 10-1, start = 000000010000000000000065, stop = 000000010000000000000065
2019-06-17 15:16:52.458 P00 DETAIL: archive retention on backup 20190617-150157F_20190617-151600D, archiveId = 10-1, start = 000000010000000000000068
2019-06-17 15:16:54.061 P00 DETAIL: remove archive: archiveId = 10-1, start = 000000010000000000000066, stop = 000000010000000000000067
2019-06-17 15:16:54.061 P00   INFO: expire command end: completed successfully (5121ms)
```

Taking additional diff backup will result in the expiring of the WAL files between the latest diff and the previous diff backup:

Taking the diff backup after adding some data and switching the WAL:

```
2019-06-17 15:29:35.523 P00   INFO: backup command begin 2.10: --no-backup-standby --compress-level=1 --config=/conf/postgresql/pgbackrest.conf --log-level-console=info --log-level-file=off --log-level-stderr=warn --pg1-path=/data/postgresql/10 --pg1-port=5432 --process-max=2 --repo1-path=/postgresql_backup_melrifa2-pg-06-16-10-53 --repo1-retention-diff=2000 --repo1-retention-full=2000 --repo1-s3-bucket=melrifa2-pg-06-16-10-53 --repo1-s3-endpoint=s3-api.us-geo.objectstorage.service.networklayer.com --repo1-s3-key=<redacted> --repo1-s3-key-secret=<redacted> --repo1-s3-region=us-standard --repo1-type=s3 --stanza=formation --start-fast --type=diff
2019-06-17 15:29:38.589 P00   INFO: last backup label = 20190617-150157F, version = 2.10
2019-06-17 15:29:41.133 P00   INFO: execute non-exclusive pg_start_backup() with label "pgBackRest backup started at 2019-06-17 15:29:35": backup begins after the requested immediate checkpoint completes
2019-06-17 15:29:41.334 P00   INFO: backup start archive = 00000001000000000000006B, lsn = 0/6B000028
2019-06-17 15:29:45.170 P01   INFO: backup file /data/postgresql/10/pg_xact/0000 (8KB, 24%) checksum 3c8b4c3e4820703a4d5f87fa72ae9955fea93239
2019-06-17 15:29:45.256 P02   INFO: backup file /data/postgresql/10/pg_commit_ts/0000 (8KB, 49%) checksum afb3f5ae1e7b42532d664e3500ba45a0eecdb801
2019-06-17 15:29:46.040 P01   INFO: backup file /data/postgresql/10/global/pg_control (8KB, 74%) checksum 7147d476f29f13daccdc3b6d2dbc44610594da84
2019-06-17 15:29:46.137 P02   INFO: backup file /data/postgresql/10/base/16471/16499 (8KB, 99%) checksum fa11a0094d0cae15568ef4212f65c33675e98ee0
2019-06-17 15:29:46.613 P01   INFO: backup file /data/postgresql/10/pg_logical/replorigin_checkpoint (8B, 100%) checksum 347fc8f2df71bd4436e38bd1516ccd7ea0d46532
2019-06-17 15:29:46.645 P00   INFO: diff backup size = 32KB
2019-06-17 15:29:46.645 P00   INFO: execute non-exclusive pg_stop_backup() and wait for all WAL segments to archive
2019-06-17 15:29:46.748 P00   INFO: backup stop archive = 00000001000000000000006B, lsn = 0/6B000130
2019-06-17 15:29:50.903 P00   INFO: new backup label = 20190617-150157F_20190617-152935D
2019-06-17 15:29:53.852 P00   INFO: backup command end: completed successfully (18330ms)
2019-06-17 15:29:53.852 P00   INFO: expire command begin
2019-06-17 15:29:57.388 P00   INFO: full backup total < 2000 - using oldest full backup for 10-1 archive retention
2019-06-17 15:29:58.271 P00   INFO: expire command end: completed successfully (4419ms)
```

Current backup info:

```
stanza: formation
    status: ok
    cipher: none

    db (current)
        wal archive min/max (10-1): none present

        full backup: 20190617-150157F
            timestamp start/stop: 2019-06-17 15:01:57 / 2019-06-17 15:08:13
            wal start/stop: 000000010000000000000065 / 000000010000000000000065
            database size: 30.1MB, backup size: 30.1MB
            repository size: 4MB, repository backup size: 4MB

        diff backup: 20190617-150157F_20190617-151600D
            timestamp start/stop: 2019-06-17 15:16:00 / 2019-06-17 15:16:14
            wal start/stop: 000000010000000000000068 / 000000010000000000000068
            database size: 30.1MB, backup size: 32.2KB
            repository size: 4MB, repository backup size: 1.4KB
            backup reference list: 20190617-150157F

        diff backup: 20190617-150157F_20190617-152935D
            timestamp start/stop: 2019-06-17 15:29:35 / 2019-06-17 15:29:49
            wal start/stop: 00000001000000000000006B / 00000001000000000000006B
            database size: 30.1MB, backup size: 32.2KB
            repository size: 4MB, repository backup size: 1.4KB
            backup reference list: 20190617-150157F```

```

The logs from rotation:

```
2019-06-17 15:32:02.754 P00   INFO: expire command begin 2.10: --config=/conf/postgresql//pgbackrest.conf --log-level-console=detail --log-level-file=off --log-level-stderr=warn --repo1-path=/postgresql_backup_melrifa2-pg-06-16-10-53 --repo1-retention-archive=7 --repo1-retention-archive-type=diff --repo1-retention-diff=2000 --repo1-retention-full=6 --repo1-s3-bucket=melrifa2-pg-06-16-10-53 --repo1-s3-endpoint=s3-api.us-geo.objectstorage.service.networklayer.com --repo1-s3-key=<redacted> --repo1-s3-key-secret=<redacted> --repo1-s3-region=us-standard --repo1-type=s3 --stanza=formation
2019-06-17 15:32:07.062 P00 DETAIL: archive retention on backup 20190617-150157F, archiveId = 10-1, start = 000000010000000000000065, stop = 000000010000000000000065
2019-06-17 15:32:07.062 P00 DETAIL: archive retention on backup 20190617-150157F_20190617-151600D, archiveId = 10-1, start = 000000010000000000000068, stop = 000000010000000000000068
2019-06-17 15:32:07.063 P00 DETAIL: archive retention on backup 20190617-150157F_20190617-152935D, archiveId = 10-1, start = 00000001000000000000006B
2019-06-17 15:32:08.728 P00 DETAIL: remove archive: archiveId = 10-1, start = 000000010000000000000069, stop = 00000001000000000000006A
2019-06-17 15:32:08.729 P00   INFO: expire command end: completed successfully (5977ms)
```

**Comment by dwsteele** [comment]
Well, this certainly looks odd.  Your settings are unusual (retention-diff=2000) but I don't see why it's not working.

We have recently migrated the expire command to C ( though it is not committed yet) so I'm not going to look at the old Perl code, but I will test your case against the new code and see if it works.  I'm planning to review that code in the next few days and I'll let you know the results.

**Comment by melrifa1** [comment]
Thanks @dwsteele I am looking forward for the results. We have `retention-diff=2000` is set this way to prevent auto backup expiry for diff backups.

**Comment by dwsteele** [comment]
I have verified this is an issue in 2.14.  The good news is that the new C code does not exhibit the same problem.  We haven't committed that yet but we expect to do that in the next day or so.

**Comment by melrifa1** [comment]
Thanks for confirmation @dwsteele

**Comment by dwsteele** [comment]
This has been fixed at 434cd832855e4e189.  How would you like your name to appear in the release notes?

**Comment by melrifa1** [comment]
We should expect the fix in 2.15 release which is scheduled to release when? My name is Mohamad El-Rifai

**Comment by dwsteele** [comment]
The plan is to release 2.15 on Monday, but we have a few things left to do so it's not definite.

Release notes added here 62715ebf2d8.
